File: app/src/main/java/org/gittner/osmbugs/parser/OsmNotesParser.java
package org.gittner.osmbugs.parser;

import org.gittner.osmbugs.bugs.OsmNote;
import org.gittner.osmbugs.common.OsmNoteComment;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

/* Parser for Openstreetmap Notes bug lists retrieved from notes */
public class OsmNotesParser
{
    public static ArrayList<OsmNote> parse(InputStream stream)
    {
        ArrayList<OsmNote> bugs = new ArrayList<>();
        try
        {
            Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(stream);
            doc.getDocumentElement().normalize();

            NodeList nList = doc.getElementsByTagName("note");
            for (int i = 0; i != nList.getLength(); ++i)
            {
                Element wpt = (Element) nList.item(i);
                double lat = Double.parseDouble(wpt.getAttribute("lat"));
                double lon = Double.parseDouble(wpt.getAttribute("lon"));

                OsmNote.STATE state = OsmNote.STATE.CLOSED;
                if (wpt.getElementsByTagName("status").item(0).getTextContent().equals("open"))
                {
                    state = OsmNote.STATE.OPEN;
                }

                long id = Long.parseLong(wpt.getElementsByTagName("id").item(0).getTextContent());

                NodeList nListComments = wpt.getElementsByTagName("comment");
                ArrayList<OsmNoteComment> comments = new ArrayList<>();
                for (int n = 0; n != nListComments.getLength(); ++n)
                {
                    OsmNoteComment comment = new OsmNoteComment();
                    comment.setText(((Element) nListComments.item(n)).getElementsByTagName("text").item(0).getTextContent());

                    NodeList element = ((Element) nListComments.item(n)).getElementsByTagName("user");
                    if (element.getLength() != 0)
                    {
                        comment.setUsername(element.item(0).getTextContent());
                    }

                    comment.setDate(((Element) nListComments.item(n)).getElementsByTagName("date").item(0).getTextContent());

                    comments.add(comment);
                }

                String text = "";
                String username = "";
                String creationDate = "";

                /* The first comment is the Bugs main Info (Description, date and user */
                if (comments.size() > 0)
                {
                    username = comments.get(0).getUsername();
                    creationDate = comments.get(0).getDate();
                    text = comments.get(0).getText();
                    comments.remove(0);
                }

                bugs.add(new OsmNote(lat, lon, id, text, username, creationDate, comments, state));
            }
        }
        catch (ParserConfigurationException | IOException | SAXException e)
        {
            e.printStackTrace();
        }
        return bugs;
    }
}

File: app/src/main/java/org/gittner/osmbugs/osmnotes/OsmNotesParser.kt
package org.gittner.osmbugs.osmnotes

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.gittner.osmbugs.osmnotes.OsmNote.OsmNoteComment
import org.joda.time.format.DateTimeFormat
import org.osmdroid.util.GeoPoint
import org.w3c.dom.Element
import org.w3c.dom.NodeList
import org.xml.sax.InputSource
import java.io.StringReader
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory


// Parser for Keepright bug lists retrieved from points.php
class OsmNotesParser {
    suspend fun parse(data: String): ArrayList<OsmNote> = withContext(Dispatchers.Default) {
        val bugs: ArrayList<OsmNote> = ArrayList()

        val doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(InputSource(StringReader(data)))
        doc.documentElement.normalize()

        val formatter = DateTimeFormat.forPattern("YYYY-MM-dd HH:mm:ss zzz")

        val nList = doc.getElementsByTagName("note")
        for (i in 0 until nList.length) {
            val note = nList.item(i) as Element

            val lat = note.getAttribute("lat").toDouble()
            val lon = note.getAttribute("lon").toDouble()

            val state = if (note.getElementsByTagName("status").item(0).textContent == "open") OsmNote.STATE.OPEN else OsmNote.STATE.CLOSED

            val id = note.getElementsByTagName("id").item(0).textContent.toLong()
            val nListComments = note.getElementsByTagName("comment")

            val date = formatter.parseDateTime(note.getElementsByTagName("date").item(0).textContent)

            val comments = ArrayList<OsmNoteComment>()
            for (n in 0 until nListComments.length) {
                val text = ((nListComments.item(n) as Element).getElementsByTagName("text").item(0).textContent)

                val creationDate = formatter.parseDateTime((nListComments.item(n) as Element).getElementsByTagName("date").item(0).textContent)

                val element: NodeList = (nListComments.item(n) as Element).getElementsByTagName("user")
                val username = if (element.length != 0) element.item(0).textContent else null

                comments.add(OsmNoteComment(text, creationDate, username ?: ""))
            }

            // The first comment is the Bugs main Info (Description, date and user)
            val firstComment = comments[0]
            comments.removeAt(0)

            bugs.add(OsmNote(GeoPoint(lat, lon), id, firstComment.Comment, firstComment.User!!, date, state, comments))
        }

        bugs
    }
}